flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationITCase.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.test.checkpointing;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.runtime.state.AbstractStateBackend;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.testutils.junit.SharedReference;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * This verifies that checkpointing works correctly for Changelog state backend with materialized
+ * state / non-materialized state.
+ */
+public class ChangelogPeriodicMaterializationITCase
+        extends ChangelogPeriodicMaterializationTestBase {
+
+    public ChangelogPeriodicMaterializationITCase(AbstractStateBackend delegatedStateBackend) {
+        super(delegatedStateBackend);
+    }
+
+    @Before
+    public void setup() throws Exception {
+        super.setup();
+    }
+
+    /** Recovery from checkpoint only containing non-materialized state. */
+    @Test
+    public void testNonMaterialization() throws Exception {
+        StreamExecutionEnvironment env =
+                getEnv(new DelegatedStateBackendWrapper(delegatedStateBackend, t -> t));
+        waitAndAssert(
+                buildJobGraph(
+                        env,
+                        new ControlledSource() {
+                            @Override
+                            protected void beforeElement(SourceContext<Integer> ctx)
+                                    throws Exception {
+                                if (getRuntimeContext().getAttemptNumber() == 0
+                                        && currentIndex == TOTAL_ELEMENTS / 2) {
+                                    waitWhile(() -> completedCheckpointNum.get() <= 0);
+                                    throwArtificialFailure();
+                                }
```

Comment:
Is it guaranteed that there was no materialization? (or it's not a strict requirement?)

Reply:
I have set `materializationInterval` to `Long.MAX_VALUE` and added a `Preconditions.checkState` to make sure there was no materialization.